src/Giraffe/HttpHandlers.fs

```diff
@@ -43,17 +45,34 @@ let private getPath (ctx : HttpContext) =
     | Some p -> ctx.Request.Path.ToString().[p.Length..]
     | None   -> ctx.Request.Path.ToString()
 
+// let private handlerWithRootedPath (path : string) (handler : HttpHandler) : HttpHandler =
+//     fun (next : HttpFunc) (ctx : HttpContext) ->
+//         task {
+//             let savedSubPath = getSavedSubPath ctx
+//             try
+//                 ctx.Items.Item RouteKey <- ((savedSubPath |> Option.defaultValue "") + path)
+//                 return! handler next ctx
+//             finally
+//                 printfn "saved subpath finalising is: %A on path %s" savedSubPath path 
+//                 match savedSubPath with
+//                 | Some savedSubPath -> ctx.Items.Item   RouteKey <- savedSubPath.Substring(0,savedSubPath.Length - path.Length)
+//                 | None              -> ctx.Items.Remove RouteKey |> ignore
+//         }
+
 let private handlerWithRootedPath (path : string) (handler : HttpHandler) : HttpHandler =
     fun (next : HttpFunc) (ctx : HttpContext) ->
-        async {
+        task {
             let savedSubPath = getSavedSubPath ctx
-            try
-                ctx.Items.Item RouteKey <- ((savedSubPath |> Option.defaultValue "") + path)
-                return! handler next ctx
-            finally
+            ctx.Items.Item RouteKey <- ((savedSubPath |> Option.defaultValue "") + path)
+            let rtask = handler next ctx
+            let! result = rtask
+            match result with
+            | Some _ -> () 
+            | None ->
                 match savedSubPath with
-                | Some savedSubPath -> ctx.Items.Item RouteKey <- savedSubPath
+                | Some savedSubPath -> ctx.Items.Item   RouteKey <- savedSubPath.Substring(0,savedSubPath.Length - path.Length)
                 | None              -> ctx.Items.Remove RouteKey |> ignore
+            return! rtask
         }
 
```

Comment:
Apologies, my bad, I am getting logic mixed up with my router that maintains the state in the dictionary, whereas here we keep a copy of string to re-insert after.

Before, we always reset the savedSubPath, whereas above, we only reset savedSubPath if route failed so as to allow further matching.

I updated to revert back and bizarrely am now failing on a completely unrelated test
```GET "/foo/blah blah/bar" returns "blah blah"`` (`HttpHandlerTests.fs` Line 453)
Looking over the test, I'm not sure why it is expecting `blah%20blah` as the result when given `blah blah` as it is usually the browser (& server) that substitute the space for `%20` and would not expect NSubstitiute dummy to replicate!? is there middleware or internal function that should be doing this substitution on path before/while parsing?